integrations: Branch filtering and related macros.

Niloth-p opened the pull request:
Fixes: #33736.
Fixes: #33738. 

Possible Follow-up: We could consider separately adding support for `*` as a wildcard character in the branch-filtering text input.

<details>
<summary><h3>Screenshots of the UI</h3></summary>

![image](https://github.com/user-attachments/assets/faf44b59-9820-4855-a931-165c6338aef4)
![image](https://github.com/user-attachments/assets/fe1db37e-1402-4e2d-a29a-a427d30b6f96)

</details>

<details>
<summary><h4>Screen capture of the UI</h4></summary>

https://github.com/user-attachments/assets/13268473-fd96-46aa-b3cf-f9bbd2ff35a7

</details>

<details>
<summary><h4>Doc screenshots</h4></summary>

https://zulip.com/integrations/doc/azuredevops
![image](https://github.com/user-attachments/assets/bbf765bc-df9f-4923-b71c-1010bba0f746)

https://zulip.com/integrations/doc/bitbucket2
![image](https://github.com/user-attachments/assets/b0666b7b-a66f-4313-93b8-21a96a1e85bb)

https://zulip.com/integrations/doc/bitbucket3
![image](https://github.com/user-attachments/assets/cf652685-9483-45ef-80ef-99c25ddec082)

https://zulip.com/integrations/doc/gitea
![image](https://github.com/user-attachments/assets/ea1b9266-87f0-4d3b-a2dc-0301db4b6bf0)

https://zulip.com/integrations/doc/gogs
![image](https://github.com/user-attachments/assets/c3544f97-4e9a-4d36-977e-dd9cb0369a45)

https://zulip.com/integrations/doc/rhodecode
![image](https://github.com/user-attachments/assets/849f526e-3fac-46aa-aa99-b8c161dea803)

https://zulip.com/integrations/doc/gitlab
![image](https://github.com/user-attachments/assets/198f4037-57bf-43f2-9f23-14e6cf1c42b0)

https://zulip.com/integrations/doc/beanstalk
![image](https://github.com/user-attachments/assets/7a84e979-38e4-4cf8-a7d1-d2b25c2e3933)

https://zulip.com/integrations/doc/bitbucket
![image](https://github.com/user-attachments/assets/32cbf7fc-d4cc-46ed-bab6-ddf8f6ebdbe1)

</details>

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented on zerver/webhooks/azuredevops/doc.md:
```diff
@@ -28,10 +28,6 @@ Get Azure DevOps notifications in Zulip!
 
 {!event-filtering-additional-feature.md!}
 
-### Configuration options
```
Approaches considered for updating the docs:
- Edit them as done in #33737, removing the macro. Since all of these except the GitHub doc use the exact same lines, considered adding a new macro with that 1 line appended to the generate-url macro instruction.
- But decided to leave it out, because "You'll be able to configure which branches you'll receive notifications from" is less of an instruction than something like "Uncheck **Send notifications for all branches**, and enter a comma-separated list of branches that you want to receive notifications" which we would have included if it was in `/help/generate-integration-url`. But, we can't add that to the help doc, since we can't selectively show that for the integrations with branch filtering available.

alya replied:
> Add a new macro for the edit. Since all of these integrations except the GitHub doc use the exact same lines, we could add a new macro with that 1 line appended to the generate-url macro instruction.

Yes, this sounds like the right approach.

alya commented:
The italics in the issue description were intentional.

Also, let's tweak the text to match the checkbox option better:

> Which branches should notifications be sent for? (comma-separated list)

alya commented:
We should probably put your cursor in the text box with the branches when you uncheck the checkbox.

alya commented:
The rest of the UI changes look fine to me, but I haven't tested yet.

Niloth-p commented:
> The italics in the issue description were intentional.

Oops, overlooked that. Fixed that now.

> put your cursor in the text box with the branches when you uncheck the checkbox

Good catch, thank you, updated.

> that should go into a helper in `zerver/lib/webhooks/git.py`

Great tip, thank you, added a function, and re-used it.

<details>
<summary><h3>Updated screenshots</h3></summary>

![image](https://github.com/user-attachments/assets/693e9b08-e810-41e2-aec2-99deb49c0af7)
![image](https://github.com/user-attachments/assets/364b1da5-a132-4c1f-8258-0a28951d5921)

</details>

alya commented:
Hm, can we put the cursor *after* `main`?

alya commented:
Tested manually, and everything else seems fine other than those two points. We can consider shortening the field label to make it fit on one line, but I think clarity is more important here.

alya commented:
By the way, I filed #33788 while looking at this.

timabbott commented:
I bet we can remove the "comma-separated list" text by making it a simple pill widget using `input_pill.ts` for inputting branches. 

Mentioning that possibility because it might be pretty fast and might naturally remove some UX issues; at least `email_pill.ts` is only 50 lines of code, and this seems even simpler.

Niloth-p commented:
Sorry for the delay.

Just added a single `place_caret_at_end` line. I noticed the same, but in my haste, I just ran with what I had.

<details>
<summary><h4>Updated Screenshot</h4></summary>

![image](https://github.com/user-attachments/assets/b743b666-a1db-4b13-a8e7-00f97f75a44b)

</details>

#### The Width

It uses the width set for the class `modal_text_input` used for text inputs in different modals. In this modal, the Topic Input  and the config-options text inputs use the same `modal_text_input` class, and would be aligned with the branch-filtering input box.
I didn't look into the width or any other class properties, I just copy-pasted the CSS class name from other text input elements.

The user profile modal seems to use some specific styling that it uses for all of its fields, while the integration-URL modal styles each component independently. So, matching the profile-modal look for the current generate-integration-URL modal would require changing multiple parts of the modal UI, and not just this new branch-filtering part.
I'm not familiar with the UI conventions yet, I'll have to look into it.

#### Pill Widget

I haven't yet started to look into this. Worst case, I might get to this after sleep, so I've updated the branch for the moment.
Will update ASAP though, because that would definitely be more user-friendly.

timabbott commented:
Let's merge this checkpoint. I think that modal width thing should be a discussion in #frontend, since it seems we are using the standard for that component, and it might be a bit of a rabbit hole. 

And since you point out we might still need that sort of text with the pill widget, we consider that in a follow-up too.

alya commented:
OK, I dunno, we can leave the modal with question for the redesign pass we plan to do on all the modals.
